Add git-merge-range and merge-range strategy

guludo opened the pull request:
This PR adds a new command called `git-merge-range` and makes use of as a new git-pile-am strategy named `merge-range`. The following paragraphs are extracted from f3c1895e25b8 ("am: Introduce the merge-range strategy") to provide the motivation:

```
It is not uncommon to have conflicts when applying a large series to the
pile (like a rebase onto a new baseline) after some development from
other contributors has been incorporated since the creation of the
series.

Using git-merge-range has proven to be very useful in those situations,
because it reduces the number of conflicts that one would run into when
applying the series and, if conflicts happen, they are arguably easier
to work with, since we handle them at the "file level" instead of
working directly on the patches.

While one can do the steps to use git-merge-range manually, it is much
better to the user that the "am" operation automates that by having a
"merge-range" strategy, so let's implement it.
```

guludo committed:
git-merge-range: Add command

This is a useful command to perform merges of revision ranges using a
strategy based on the 3-way merge of files.

In multiple situations it was proven very useful to re-apply a pile
series, avoiding conflicts that would arise otherwise. In an upcoming
change, a new strategy for `git pile am` will be added to allow using
this, where details of how this is used for pile will be provided.

Note that this is kept as a separate "self-contained" directory because
this can be a separate command of its own, independent of git-pile. In
the future we might try incorporating it into git.

The current implementation is quite lazy and uses git-range-diff to find
matching commits between ranges instead of implementing its own logic.
As we are currently only interested in range-diff lines with "=" type of
matches, we could get away with an algorithm not so time-complex as the
one used by git-range-diff. We can do such optimization in the future.

Signed-off-by: Gustavo Sousa <gustavo.sousa@intel.com>

guludo committed:
am: Introduce the merge-range strategy

It is not uncommon to have conflicts when applying a large series to the
pile (like a rebase onto a new baseline) after some development from
other contributors has been incorporated since the creation of the
series.

Using git-merge-range has proven to be very useful in those situations,
because it reduces the number of conflicts that one would run into when
applying the series and, if conflicts happen, they are arguably easier
to work with, since we handle them at the "file level" instead of
working directly on the patches.

While one can do the steps to use git-merge-range manually, it is much
better to the user that the "am" operation automates that by having a
"merge-range" strategy, so let's implement it.

It would be nice to have all tests in 60_format_patch_and_am.bats to be
parametrized by the strategy, so that we have more coverage. However, an
attempt to use bats_test_function as in [1] failed with bats 1.10.0.
That's why, for now, we fork only test format-patch-and-am-strategy, so
that there is one for the "top" strategy and another for the
"merge-range" strategy.

[1] https://github.com/bats-core/bats-core/blob/990d8e285e7564ff61ea2c93a6da12856d337d83/test/fixtures/bats/dynamic_test_registration.bats

Signed-off-by: Gustavo Sousa <gustavo.sousa@intel.com>

guludo committed:
gitutil: Try to remove temporary workdir first

We the current state of git_temporary_worktree(), probably due to
timing, sometimes we would end up with a failure when removing the
worktree, most likely because the cleanup performed TemporaryDirectory
took place in the file system.

Let's invert stuff here by having TemporaryDirectory wrapping the try
statement so that we gracefully remove the worktree before the cleanup
operation kicks in.

Note that with this change, we also need to use the --force flag to the
worktree removal command, as there might be unstaged/untracked files in
it.

Signed-off-by: Gustavo Sousa <gustavo.sousa@intel.com>

guludo committed:
git-merge-range: Use own logic for matching commits

Using range-diff to find matching commits is overkill, since we are only
interested in "full" ("=") matches and not the other cases ("!", ">",
"<"). As such, implement our own commit matching logic. With this
change, we observed a speedup of about 2.28 on a workload where:

  - Current, base and other ranges contained 414, 424 and 410
    commits, respectively;
  - Most of the commits matched.

Signed-off-by: Gustavo Sousa <gustavo.sousa@intel.com>

guludo commented:
Force-pushed to make linters happy.

lucasdemarchi commented:
Humn... tests currently failing and it seems unrelated to previous CI breakage:

https://github.com/git-pile/git-pile/actions/runs/10724118311#summary-29739028102

guludo commented:
Looks like we are getting non-zero exit status for the command `git var GIT_SEQUENCE_EDITOR`. I guess I'll need to run the test on the same container to debug this.

Another error specific for git v2.20.* is that we are getting non-zero result from `git branch --show-current`. Maybe because that flag is not supported on that version.

guludo commented:
> Looks like we are getting non-zero exit status for the command `git var GIT_SEQUENCE_EDITOR`. I guess I'll need to run the test on the same container to debug this.

Yeah. I confirmed that this command simply fails in the container. The support for `GIT_SEQUENCE_EDITOR` in `git var` was added with https://github.com/git/git/commit/4c3dd9304e49402bd4ee19dfaa4c21d0217fb582, which seems to be present starting with git v2.40.

As a fallback, we could implement a python function that mimics what git does: https://github.com/git/git/blob/2e7b89e038c0c888acf61f1b4ee5a43d4dd5e94c/editor.c#L48

```c
const char *git_sequence_editor(void)
{
	const char *editor = getenv("GIT_SEQUENCE_EDITOR");

	if (!editor)
		git_config_get_string_tmp("sequence.editor", &editor);
	if (!editor)
		editor = git_editor();

	return editor;
}
```

lucasdemarchi commented:
> As a fallback, we could implement a python function that mimics what git does: https://github.com/git/git/blob/2e7b89e038c0c888acf61f1b4ee5a43d4dd5e94c/editor.c#L48

sounds good to me. Probably good to just use `git var GIT_SEQUENCE_EDITOR` and fallback on error?

guludo committed:
git-merge-range: Add fallback for git var GIT_SEQUENCE_EDITOR

Support for GIT_SEQUENCE_EDITOR in the command "git var" was added only
with git v2.40. Let's mimic git's implementation in case the call to
"git var GIT_SEQUENCE_EDITOR" fails.

Signed-off-by: Gustavo Sousa <gustavo.sousa@intel.com>

guludo commented:
> > As a fallback, we could implement a python function that mimics what git does: https://github.com/git/git/blob/2e7b89e038c0c888acf61f1b4ee5a43d4dd5e94c/editor.c#L48
> 
> sounds good to me. Probably good to just use `git var GIT_SEQUENCE_EDITOR` and fallback on error?

Yep. That's what I ended up doing.

guludo commented:
> > > As a fallback, we could implement a python function that mimics what git does: https://github.com/git/git/blob/2e7b89e038c0c888acf61f1b4ee5a43d4dd5e94c/editor.c#L48
> > 
> > 
> > sounds good to me. Probably good to just use `git var GIT_SEQUENCE_EDITOR` and fallback on error?
> 
> Yep. That's what I ended up doing.

And we will probably need to do something similar for `git branch --show-current`.

lucasdemarchi commented on git_merge_range/args.py:
```diff
@@ -0,0 +1,157 @@
+Each "current", "base" and "other" argument can be either a revision range or a
+single revision. In the case of the latter, the single revision is taken as the
+head and the merge base (git merge-base --octopus ...) of the three ranges'
+heads is used as default base.
```
I'm more a example-driven person and like the ascii art examples in git-log(1) / git-merge(1). Could you add one for this command?

guludo replied:
Yep. I'll try to do it when I get some time.

guludo commented:
@lucasdemarchi, looks like there are some more environment-related CI issues to solve here. Since I don't have the time to work on them right now, I do not want to block the next release of git-pile on this.

lucasdemarchi commented:
@guludo ok... doing a release is very cheap. We can release 1.2, then fix this without rushing it and then doing a 1.3.

Commit:
Merge branch 'master' into merge-range
